src/main/java/org/jenkinsci/plugins/credentialsbinding/MultiBinding.java

```diff
@@ -80,21 +80,37 @@ public final String getCredentialsId() {
     public static final class MultiEnvironment implements Serializable {
 
         private final Map<String,String> values;
+        private final Map<String,String> secretValues;
         private final Unbinder unbinder;
 
-        public MultiEnvironment(Map<String,String> values) {
-            this(values, new NullUnbinder());
+        public MultiEnvironment(Map<String,String> secretValues) {
+            this(secretValues, Collections.<String, String>emptyMap());
         }
 
-        public MultiEnvironment(Map<String,String> values, Unbinder unbinder) {
-            this.values = new HashMap<String,String>(values);
+        public MultiEnvironment(Map<String,String> secretValues, Map<String,String> publicValues) {
+            this(secretValues, publicValues, new NullUnbinder());
+        }
+
+        public MultiEnvironment(Map<String,String> secretValues, Unbinder unbinder) {
+            this(secretValues, Collections.<String, String>emptyMap(), unbinder);
+        }
+
+        public MultiEnvironment(Map<String,String> secretValues, Map<String,String> publicValues, Unbinder unbinder) {
+            this.secretValues = new HashMap<String,String>(secretValues);
+            this.values = new HashMap<String,String>();
+            this.values.putAll(secretValues);
```

Comment:
So we store two copies of the secret values? That does not seem right.

I would suggest deprecating the `values` field and introducing a `publicValues` field; the constructor would initialize `secretValues` and `publicValues`, with `readResolve` doing the migration for the corner case.